packages/harmony/src/components/input/TextInput/TextInput.tsx

```diff
@@ -0,0 +1,213 @@
+   */
+  startIcon?: IconComponent
+  /**
+   * Floating text on the righthand side of the input
+   */
+  endIcon?: IconComponent
+  /**
+   * Required or not. Will add an * to the label if required
+   */
+  required?: boolean
+}
+
+export const TextInput = forwardRef<HTMLInputElement, TextInputProps>(
+  (props: TextInputProps, ref) => {
+    const {
+      required,
+      label: labelProp,
+      className,
+      inputRootClassName,
+      maxLength,
+      showMaxLength,
+      size = TextInputSize.DEFAULT,
+      hideLabel,
+      value,
+      children,
+      warning: warningProp,
+      error,
+      className: inputClassName,
+      disabled,
+      onFocus: onFocusProp,
+      onBlur: onBlurProp,
+      placeholder,
+      helperText,
+      startAdornmentText,
+      endAdornmentText,
+      startIcon: StartIcon,
+      endIcon: EndIcon,
+      ...other
+    } = props
+
+    const characterCount = value !== undefined ? `${value}`.length : 0
+    const nearCharacterLimit = maxLength && characterCount >= 0.9 * maxLength
+
+    // Hide the label when: requested, when the size is small, or when adornment text is present
+    const shouldShowLabel =
+      !hideLabel &&
+      size !== TextInputSize.SMALL &&
+      endAdornmentText === undefined &&
+      startAdornmentText === undefined
+
+    const label = required ? `${labelProp} *` : labelProp
+
+    /**
+     * Since Firefox doesn't support the :has() pseudo selector,
+     * manually track the focused state and use classes for focus, required, and disabled
+     */
+    const [isFocused, handleFocus, handleBlur] = useFocusState(
+      onFocusProp,
+      onBlurProp
+    )
+
+    // Whenever a label isn't visible the placeholder shows in it's place
+    const shouldShowPlaceholder = isFocused || !shouldShowLabel
+
+    const style = {
+      [styles.default]: size === TextInputSize.DEFAULT,
+      [styles.small]: size === TextInputSize.SMALL,
+      [styles.warning]: warningProp || nearCharacterLimit,
+      [styles.error]: error,
+      [styles.focused]: isFocused,
+      [styles.disabled]: disabled,
+      [styles.required]: required
+    }
+
+    const inputRender = (
+      <div className={cn(styles.inputRow, layoutStyles.row)}>
+        <div className={cn(layoutStyles.row, styles.inputContainer)}>
+          {startAdornmentText ? (
+            <Text variant='label' size='l' color='subdued'>
+              {startAdornmentText}
+            </Text>
+          ) : null}
+          <input
+            onFocus={handleFocus}
+            onBlur={handleBlur}
+            ref={ref}
+            className={cn(styles.textInput, inputClassName)}
+            value={value}
+            maxLength={maxLength}
+            disabled={disabled}
+            placeholder={shouldShowPlaceholder ? placeholder : undefined}
+            aria-label={
+              label ?? endAdornmentText ?? startAdornmentText ?? placeholder
```

Comment:
are the adornments more appropriate than the placeholder? maybe `label ?? startAdornment + placeholder + endAdornment`?

Reply:
Changed this behavior, see my other comment https://github.com/AudiusProject/audius-protocol/pull/6329#discussion_r1364280814